### ISSUE TITLE
Dependency generation issue

### ISSUE THREAD

**Issue by bendudson** [boutproject/BOUT-dev#997]
This line: https://github.com/boutproject/BOUT-dev/blob/next/make.config.in#L306
leads to an error when compiling on the YPI machines:

```
  Compiling  nc_format.cxx
make[4]: m2c: Command not found
<builtin>: recipe for target '/opt/york/phys/pkg/Modules/netcdf/4.4.1.1/1/gcc-5.4.0/include/netcdf' failed
make[4]: *** [/opt/york/phys/pkg/Modules/netcdf/4.4.1.1/1/gcc-5.4.0/include/netcdf] Error 127
../../../make.config:115: recipe for target 'netcdf4' failed
make[3]: *** [netcdf4] Error 2
../../make.config:115: recipe for target 'impls' failed
make[2]: *** [impls] Error 2
```

This thread (https://lists.mcs.anl.gov/pipermail/petsc-users/2010-November/007299.html) indicates that the problem may be an implicit rule for building modula-2 .mod files.

Some key parts of my `make.config` file are:

```
BOUT_FLAGS              = $(CXXFLAGS) -g -O2 -O -DCHECK=2 -DSIGHANDLE -DLOGCOLOR -DBOUT_VERSION_STRING="\"4.1.2\"" -DBOUT_VERSION_DOUBLE=4.12  -DBACKTRACE -DNCDF4 -DHDF5 -DLAPACK -DBOUT_HAS_PVODE  -std=c++11

CXX                     = /opt/york/phys/pkg/Modules/openmpi/1.10.7/1/gcc-5.4.0/bin/mpicxx

-include .*.mk
```
Tested on the `sawtooth` machine with :
```
Currently Loaded Modulefiles:
  1) null             3) its-hdd          5) openmpi/1.10.7   7) szip/2.1.1       9) netcdf/4.4.1.1  11) fftw/3.3.6
  2) its-net          4) gcc/5.4.0        6) zlib/1.2.11      8) hdf5/1.8.18     10) sundials/2.7.0  12) petsc/3.7.6
```
Configuring with no arguments:
```
Configuration summary

  PETSc support: no
  SLEPc support: no
  IDA support: no
  CVODE support: no
  ARKODE support: no
  NetCDF support: yes
  Parallel-NetCDF support: no
  HDF5 support: yes (parallel: no)
  MUMPS support: no
  Lapack support: yes
  Scorep support: no
```

**Comment by dschwoerer** [comment]
Assuming this is the same issue - why do you have mod files?

Assuming it is a different issue: Can you run make with: `MAKEFLAGS=-R make` - to disable build in rules.

If that works - can you run make again with `MAKEFLAGS=-d make -j1` to debug make?
That will be VERY noisy - but if you grep for m2c that might help:
`MAKEFLAGS=-d make -j1 2>&1 | grep m2c -B 100 -A 100`

**Comment by ZedThree** [comment]
I've fixed this particular case on our systems, but it hasn't solved the actual problem.

The issue is that we `#include <netcdf>`, i.e. a file called just `netcdf`. This is then put into the dependency file (`.ncxx4.mk`) as a dependency (see below). In the same include directory (`/opt/york/.../include/`), there is also a file called `netcdf.mod` (a Fortran module file) which has a time stamp newer than `netcdf`. Finally, a make implicit rule tries to build `netcdf` from `netcdf.mod` using `m2c`, which obviously fails.

Touching the `netcdf` file fixes this, as make no longer tries to build it.

I've just checked Archer, and `netcdf.mod` is older than `netcdf` there (at least for the `cray-netcdf` module), but we should be wary about other machines where this may crop up again.

An actual solution to this and #998 would be to make the dependency generation entirely optional, with a separate `make depends` target (suggested by @bendudson).

---

The dependency file looks like:

```
ncxx4.o: ncxx4.cxx ncxx4.hxx ../../../../include/dataformat.hxx \
 ../../../../include/bout_types.hxx ../../../../include/unused.hxx \
 ../../../../include/unused.hxx \
 /opt/york/phys/pkg/Modules/netcdf/4.4.1.1/1/gcc-5.4.0/include/netcdf \
...
```
